man/touch.3
.Dd December 6th, 2021
.Dt TOUCH 3L LOCAL
.Os Codella 0.8.4
.
.
.Sh NAME
.Nm touch
.Nd タッチパネル制御クラス
.
.
.Sh LIBRARY
.Lb Codella (libarmadillo, \-larmadillo)
.
.
.Sh SYNOPSIS
.In armadillo/touch.hh
.Vt namespace armadillo ;
.Vt class armadillo::touch ;
.Vt typedef struct {
.Vt 	int p , x , y ;
.Vt } armadillo::touch::touch_point_t ;
.Vt typedef struct {
.Vt 	int type ; // イベントの種別
.Vt 	int touches ; // 総タッチ数
.Vt 	int touch ; // イベントの対象であるタッチ番号
.Vt 	touch_point_t value ; // タッチの座標と強さ
.Vt } armadillo::touch::touch_event_t ;
.Fn armadillo::touch::touch
.Fn armadillo::touch::\(titouch
.Ft void
.Fn armadillo::touch::enable
.Ft void
.Fn armadillo::touch::disable
.sp
// 次の型はマニュアルで表記を簡単にするためだけのものです
.Vt typedef void (*handler_t)(armadillo::touch::touch_event_t, void *) ;
.Ft void
.Fn armadillo::touch::on_none "handler_t handler" "void *arg = NULL"
.Ft void
.Fn armadillo::touch::on_start "handler_t handler" "void *arg = NULL"
.Ft void
.Fn armadillo::touch::on_move "handler_t handler" "void *arg = NULL"
.Ft void
.Fn armadillo::touch::on_end "handler_t handler" "void *arg = NULL"
.
.
.Sh DESCRIPTION
名前空間
.Vt armadillo
に配置されるクラス
.Vt touch
を利用してタッチパネルを制御できます.
インスタンスを生成するとデバイスは利用可能な状態となります.
インスタンスが破棄されるとデバイスは解放されます.
.Pp
コンストラクタ
.Fn touch
はデバイスと紐付けられたインスタンスを生成します.
.Pp
デストラクタ
.Fn \(titouch
はインスタンスに紐付けられたデバイスを解放します.
.Pp
メンバ関数
.Fn enable
はタッチパネルを有効にします.
.Pp
メンバ関数
.Fn disable
はタッチパネルを無効にします.
この関数を呼び出した後にメンバ関数
.Fn enable
を再度呼び出すことでセンサを再度有効にできます.
.Pp
メンバ関数
.Fn on_none
はタッチの開始や移動, 終了もない場合に発火するイベントのハンドラを
.Fa handler
に設定します.
省略可能な
.Fa arg
を指定すると
.Fa handler
が呼び出されたときの第二引数として利用できます.
.Pp
メンバ関数
.Fn on_start
はタッチの開始時に発火するイベントのハンドラを
.Fa handler
に設定します.
省略可能な
.Fa arg
を指定すると
.Fa handler
が呼び出されたときの第二引数として利用できます.
.Pp
メンバ関数
.Fn on_move
はタッチの移動時に発火するイベントのハンドラを
.Fa handler
に設定します.
省略可能な
.Fa arg
を指定すると
.Fa handler
が呼び出されたときの第二引数として利用できます.
.Pp
メンバ関数
.Fn on_end
はタッチの終了時に発火するイベントのハンドラを
.Fa handler
に設定します.
省略可能な
.Fa arg
を指定すると
.Fa handler
が呼び出されたときの第二引数として利用できます.
.Pp
イベントハンドラ
.Fa handler
は各種イベント発火時に呼び出されます.
それぞれのイベントハンドラはそれぞれのスレッドで実行されます.
第一引数はイベントの情報を含む構造体です.
構造体のメンバ
.Fa type
の値は
.Dv TE_NONE , TE_START , TE_MOVE , TE_END
のいずれかです.
.
.
.Sh IMPLEMENTATION NOTES
クラス
.Vt touch
の内部では
.Xr pthread 7
を利用しています.
同一プロセス内で複数の
.Vt touch
インスタンスが生成されると正常に動作しない恐れがあります.
.Pp
各種イベントハンドラは約 8 msec の周期で呼び出されます.
イベントハンドラの処理が長時間に渡ると,
資源不足に陥り新たなイベントハンドラを実行できなくなります.
.
.
.Sh RETURN VALUES
クラス
.Vt touch
のいかなるメンバ関数も値を返しません.
.
.
.Sh ERRORS
クラス
.Vt touch
のコンストラクタ及び各メンバ関数は,
失敗すると
.Vt std::runtime_error
の例外を発生させます.
例外のメンバ関数である
.Fn what
はエラーに関するメッセージを返します.
エラーが発生した後の
.Vt touch
インスタンスの状態は不定です.
.
.
.Sh EXAMPLES
次の例は画面に赤と青のボタンを表示し,
どちらが押下されたのか標準出力に表示します.
.Bd -literal
#include <cstdio>       // printf(3)
#include <ctime>        // nanosleep(2)
#include <armadillo/display.hh>
#include <armadillo/touch.hh>

#define B_WIDTH         200
#define B_HEIGHT        280
#define B_TOP           100
#define B_LEFT_1        100
#define B_LEFT_2        500

struct my_arg {
	armadillo::display *disp;
	struct { int x, y; } position;
	bool done;
};

void touch_start(armadillo::touch::touch_event_t ev, void *va)
{
	struct my_arg *arg = (struct my_arg *)va;

	if (ev.value.y < B_TOP || ev.value.y > B_TOP + B_HEIGHT)
		return;
	if (ev.value.x > B_LEFT_1 && ev.value.x < B_LEFT_1 + B_WIDTH) {
		arg->disp->rect(B_LEFT_1, B_TOP, B_WIDTH, B_HEIGHT,
				armadillo::display::color("red"), true);
		arg->done = true;
	}
	if (ev.value.x > B_LEFT_2 && ev.value.x < B_LEFT_2 + B_WIDTH) {
		arg->disp->rect(B_LEFT_2, B_TOP, B_WIDTH, B_HEIGHT,
				armadillo::display::color("blue"), true);
		arg->done = true;
	}

	arg->position.x = ev.value.x;
	arg->position.y = ev.value.y;
}

int main(void)
{
	armadillo::display disp;

	disp.draw_init();
	disp.rect(B_LEFT_1, B_TOP, B_WIDTH, B_HEIGHT,
			armadillo::display::color("red"));
	disp.rect(B_LEFT_2, B_TOP, B_WIDTH, B_HEIGHT,
			armadillo::display::color("blue"));
	disp.draw_end();

	struct my_arg arg;
	arg.disp = &disp;
	arg.done = false;

	armadillo::touch tp;
	tp.on_start(touch_start, &arg);

	tp.enable();
	while (arg.done != true) {
		struct timespec ts;
		ts.tv_sec  = 0;
		ts.tv_nsec = (long)1e6;
		nanosleep(&ts, NULL);
	}
	tp.disable();

	if (arg.position.x > B_LEFT_1 && arg.position.x < B_LEFT_1 + B_WIDTH)
		printf("RED\en");
	if (arg.position.x > B_LEFT_2 && arg.position.x < B_LEFT_2 + B_WIDTH)
		printf("BLUE\en");

	return 0;
}
.Ed
.Pp
このプログラムをコンパイルするには次のように実行します.
.Dl % c++ test.cc -larmadillo `pkg-config --cflags freetype2`
.
.
.Sh FILES
.Pa /dev/input/event1
.
.
.Sh COPYRIGHT
Copyright \(co 2021
.An KusaReMKN .
.Pp
プログラムは MIT License で利用可能です.
文書は CC-BY 4.0 かそれ以上で利用可能です.
.
.
./" Sh BUGS
./" とくにないとおもいます.


src/touch.cc
#include "touch.hh"

const char *const armadillo::touch::TOUCH_PATH = "/dev/input/event1";

void *armadillo::touch::handler_wrapper(void *arg)
{
	handler_arg_t *tmp = (handler_arg_t *)arg;
	touch_event_t ev = tmp->event;
	touch *self = tmp->self;
	(*self->handlers[ev.type])(ev, self->args[ev.type]);
	free(arg);
	return NULL;
}

void armadillo::touch::call_handler(handler_arg_t *arg)
{
	pthread_attr_t attr;
	pthread_attr_init(&attr);
	pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);

	int err;
	pthread_t ignore;
	do {
		err = pthread_create(&ignore, &attr, handler_wrapper, arg);
	} while (err == EAGAIN);
	if (err != 0)
		throw std::runtime_error(
			armadillo::errmsg("pthread_create", err));

	pthread_attr_destroy(&attr);
}

void *armadillo::touch::event_loop(void *me)
{
	touch *self = (touch *)me;

	int types[3][3][2] = {
		{
			{ TE_NONE,  TE_NONE  },
			{ TE_START, TE_NONE  },
			{ TE_START, TE_START }
		}, {
			{ TE_END,   TE_NONE  },
			{ TE_MOVE,  TE_NONE  },
			{ TE_MOVE,  TE_START },
		}, {
			{ TE_END,   TE_END   },
			{ TE_MOVE,  TE_END   },
			{ TE_MOVE,  TE_MOVE  },
		}
	};

	while (self->enabled) {
		self->update_frame();

		for (int i = 0; i < 2; i++) {
			int type = types[self->prev_touches][self->touches][i];
			if (self->handlers[type] == NULL)
				continue;

			handler_arg_t *arg;
			arg = (handler_arg_t *)malloc(sizeof(handler_arg_t));
			if (arg == NULL)
				throw std::runtime_error(
					armadillo::errmsg("malloc"));

			arg->self = self;
			arg->event.type = type;
			arg->event.touches = self->touches;
			arg->event.touch = i;
			arg->event.value = self->values[i];

			self->call_handler(arg);
		}
	}

	return NULL;
}

void armadillo::touch::update_frame()
{
	prev_touches = touches;
	touches = 0;

	bool sync = 0;
	bool key = 0;
	while (read_event(&sync, &key));
	if (touches == 0 && !key)
		touches = prev_touches;
}

int armadillo::touch::read_event(bool *sync, bool *key)
{
	fd_set rfs;
	FD_ZERO(&rfs);
	FD_SET(touch_file, &rfs);

	struct timeval timeout;
	timeout.tv_sec = 0;
	timeout.tv_usec = 10000;

	int err;
	err = select(touch_file + 1, &rfs, NULL, NULL, &timeout);
	if (err == -1)
		throw std::runtime_error(armadillo::errmsg("select"));
	else if (err == 0)
		return 0;

	struct input_event ev;
	err = read(touch_file, &ev, sizeof(ev));
	if (err == -1)
		throw std::runtime_error(armadillo::errmsg("read"));

	switch (ev.type) {
	case EV_ABS:
		switch (ev.code) {
		case ABS_MT_TOUCH_MAJOR:
			values[touches].p = ev.value;
			if (touches != 0)
				*sync = false;
			break;
		case ABS_MT_POSITION_X:
			values[touches].x = ev.value;
			break;
		case ABS_MT_POSITION_Y:
			values[touches].y = ev.value;
			touches++;
			break;
		case ABS_PRESSURE:
		case ABS_X:
		case ABS_Y:
			break;	// Ignore. I don't use these values.
		default:
			throw std::runtime_error(armadillo::errmsg(
					"read_event", "unknown code"));
		}
		break;
	case EV_SYN:
		if (*sync)
			return 0;
		*sync = true;
		break;
	case EV_KEY:
		*key = true;
		break;	// Ignore. I don't use this value.
	default:
		throw std::runtime_error(armadillo::errmsg("read_event",
					"unknown type"));
	}

	return 1;
}

armadillo::touch::touch()
{
	touch_file = open(TOUCH_PATH, O_RDONLY);
	if (touch_file == -1)
		throw std::runtime_error(armadillo::errmsg("open"));
	for (int i = 0; i < TE_MAX; i++) {
		handlers[i] = NULL;
		args[i] = NULL;
	}
}

armadillo::touch::~touch()
{
	close(touch_file);
}

void armadillo::touch::enable()
{
	enabled = true;
	prev_touches = touches = 0;
	std::memset(values, 0, sizeof(values));
	int err;
	err = pthread_create(&event_thread, NULL, event_loop, this);
	if (err != 0)
		throw std::runtime_error(
			armadillo::errmsg("pthread_create", err));
}

void armadillo::touch::disable()
{
	enabled = false;
	int err;
	err = pthread_join(event_thread, NULL);
	if (err != 0)
		throw std::runtime_error(
			armadillo::errmsg("pthread_join", err));
}

void armadillo::touch::on_none(void (*handler)(touch_event_t, void *),
		void *arg)
{
	handlers[TE_NONE] = handler;
	args[TE_NONE] = arg;
}

void armadillo::touch::on_start(void (*handler)(touch_event_t, void *),
		void *arg)
{
	handlers[TE_START] = handler;
	args[TE_START] = arg;
}

void armadillo::touch::on_move(void (*handler)(touch_event_t, void *),
		void *arg)
{
	handlers[TE_MOVE] = handler;
	args[TE_MOVE] = arg;
}

void armadillo::touch::on_end(void (*handler)(touch_event_t, void *),
		void *arg)
{
	handlers[TE_END] = handler;
	args[TE_END] = arg;
}
